notificationworkerlambda/src/main/scala/com/gu/notifications/worker/SenderRequestHandler.scala

```diff
@@ -68,26 +68,35 @@ trait SenderRequestHandler[C <: DeliveryClient] extends Logging {
       .evalTap(Reporting.log(s"Sending failure: "))
   } yield resp
 
-  def deliverChunkedTokens(chunkedTokenStream: Stream[IO, (ChunkedTokens, Long, Instant)]): Stream[IO, Unit] = {
-    for {
-      (chunkedTokens, sentTime, functionStartTime) <- chunkedTokenStream
-      individualNotifications = Stream.emits(chunkedTokens.toNotificationToSends).covary[IO]
-      resp <- deliverIndividualNotificationStream(individualNotifications)
-        .broadcastTo(
-          reportSuccesses(chunkedTokens, sentTime, functionStartTime),
-          cleanupFailures,
-          trackProgress(chunkedTokens.notification.id))
-    } yield resp
+  def deliverChunkedTokens(chunkedTokenStream: Stream[IO, (ChunkedTokens, Long, Instant, Int)]): Stream[IO, Unit] = {
+    chunkedTokenStream.map {
+      case (chunkedTokens, sentTime, functionStartTime, sqsMessageBatchSize) =>
+        val individualNotifications = Stream.emits(chunkedTokens.toNotificationToSends).covary[IO]
+        deliverIndividualNotificationStream(individualNotifications)
+          .broadcastTo(
+            reportSuccesses(chunkedTokens, sentTime, functionStartTime, sqsMessageBatchSize),
+            cleanupFailures,
+            trackProgress(chunkedTokens.notification.id))
+      }.parJoin(maxConcurrency)
   }
 
   def handleChunkTokens(event: SQSEvent, context: Context): Unit = {
-    val chunkedTokenStream: Stream[IO, (ChunkedTokens, Long, Instant)] = Stream.emits(event.getRecords.asScala)
+    val sqsMessageBatchSize = event.getRecords.size
+    val startTime = Instant.now
+    val totalTokensProcessed: Int = event.getRecords.asScala.map(event => NotificationParser.parseChunkedTokenEvent(event.getBody)).foldLeft(0)((acc, r) => acc + r.tokens.size)
+    val chunkedTokenStream: Stream[IO, (ChunkedTokens, Long, Instant, Int)] = Stream.emits(event.getRecords.asScala)
       .map(r => (r.getBody, r.getAttributes.getOrDefault("SentTimestamp", "0").toLong))
-      .map { case (body, sentTimestamp) => (NotificationParser.parseChunkedTokenEvent(body), sentTimestamp, Instant.now) }
+      .map { case (body, sentTimestamp) => (NotificationParser.parseChunkedTokenEvent(body), sentTimestamp, startTime, sqsMessageBatchSize) }
 
     deliverChunkedTokens(chunkedTokenStream)
       .compile
       .drain
       .unsafeRunSync()
+
+    logger.info(Map(
+      "sqsMessageBatchSize" -> sqsMessageBatchSize,
+      "totalTokensProcessed" -> totalTokensProcessed,
+      "invocation.functionProcessingRate" -> { totalTokensProcessed.toDouble / Duration.between(startTime, Instant.now).toMillis * 1000 },
```

Comment:
I thought this was important to capture, to get an understanding of how well the _entire_ invocation performs